lib/exceptions.ts

```diff
@@ -1,55 +1,107 @@
+type Message = string;
+
+interface Status {
+  status: number;
+  statusText: string;
+}
+
+interface ErrorDetails {
+  signature?: string;
+  raw?: any;
+}
+
+interface FetchErrorDetails extends Status {
+  headers: Headers;
+  body: string;
+}
+
+// Deprecated
+interface AxiosErrorDetails extends Partial<Status> {
+  originalError: Error;
+  code?: string;
+}
+
 export class SignatureValidationFailed extends Error {
-  constructor(
-    message: string,
-    public signature?: string,
-  ) {
+  public signature?: string;
+
+  constructor(message: Message, { signature }: ErrorDetails = {}) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { signature });
   }
 }
 
 export class JSONParseError extends Error {
+  public raw: any;
+
+  constructor(message: Message, { raw }: ErrorDetails = {}) {
+    super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { raw });
+  }
+}
+
+export class HTTPFetchError extends Error {
+  public status: number;
+
+  public statusText: string;
+
+  public headers: Headers;
+
+  public body: string;
+
   constructor(
-    message: string,
-    public raw: any,
+    message: Message,
+    { status, statusText, headers, body }: FetchErrorDetails,
   ) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { status, statusText, headers, body });
   }
 }
 
+/* Deprecated */
 export class RequestError extends Error {
-  constructor(
-    message: string,
-    public code: string,
-    private originalError: Error,
-  ) {
+  public code: string;
+
+  private originalError: Error;
+
+  constructor(message: Message, { code, originalError }: AxiosErrorDetails) {
     super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { code, originalError });
   }
 }
 
 export class ReadError extends Error {
-  constructor(private originalError: Error) {
-    super(originalError.message);
+  public originalError: Error;
+
+  constructor(message: Message, { originalError }: AxiosErrorDetails) {
+    super(message);
+    this.name = this.constructor.name;
+
+    Object.assign(this, { originalError });
   }
 }
 
 export class HTTPError extends Error {
+  public status: number;
```

Comment:
This error type is also deprecated, because it's only used in `http-axio.ts`. I don't want to change public field in deprecated code. Please keep `statusCode` as it is. 

(Renaming `statusCode` in `HTTPFetchError` is nice. Please keep the change.)

Reply:
I didn't think of that part! :) Thank you for the great feedback.
The basis for the naming change was as follows.

```shell
1. Use the field name provided by the response header as it is.
2. Error Classes should use the same field naming.
```

As you said, it seems good to separate the interfaces of the deprecated objects and make a difference in naming. I will apply it right away. :)